feat(hogql): action matching with hogql bytecode

mariusandra opened the pull request:
## Problem

The PR https://github.com/PostHog/posthog/pull/16189 got too big.

## Changes

Part 4 of the split. This one removes part 3 (cohort matching code), and then implements HogVM action matching in the plugin server.

It also implements the following warning on the actions page, if the HogQL bytecode couldn't be generated:

![2023-06-30 00 47 13](https://github.com/PostHog/posthog/assets/53387/54df0998-f6d9-4561-a645-32e618d753f3)

It's not the prettiest, and the saving dynamic isn't ideal... but it's functional :D.


![2023-06-22 15 55 09](https://github.com/PostHog/posthog/assets/53387/c6c5ec87-96ed-44e3-a7ca-ee1e441202df)


## How did you test this code?

I created tests for bytecode, cohort and element matching. Then created similar actions myself, and verified that I need to remove cohort matching, since it's not working. So I did. I also verified element matching works.

mariusandra committed:
feat(hogql): action matching with hogql bytecode

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
improvements

mariusandra committed:
Merge branch 'hogvm-action-matching' of github.com:PostHog/posthog into hogvm-action-matching

mariusandra committed:
fix test

mariusandra committed:
fix test 2

tiina303 commented:
@mariusandra  sorry I didn't get to comment on the previous PR earlier.
There's a longer term plan that's relevant here: https://docs.google.com/document/d/17sPsimk57wZ8OQehPn0pDWJFBboADv9xXBrnu-pfoL4
specifically we're thinking of adding base events (before any person/group/cohort processing and moving that to happen later in batches). Given that folks would like webhooks to be fast we'll want to move that to be on top of base events, so we don't have access to any person info. We might be forced to support the existing users by having those webhooks after the batching is done, but lets encourage users to not setup any new webhooks via hogQL that use persons or cohorts, not sure how to best do that though 😅  It would otherwise likely be very confusing to users why some webhooks are really slow and others fast

mariusandra commented:
Hey, thanks for the context! Few thoughts/questions below.

However first, what's your general take on replacing the existing filter/action matching system in the plugin server with a stack-based VM that evaluates a compact `bytecode` array for each action? I haven't benchmarked it, but I think this straightforward loop+switch will perform much faster than the existing action matching logic. It feels conceptually simpler (each action has code to match against, no need to know about the action step/filter/property/element business logic). Plus it's straightforward to port to any other language, in case we want to move more things over to e.g. Rust.

> @mariusandra sorry I didn't get to comment on the previous PR earlier. There's a longer term plan that's relevant here: https://docs.google.com/document/d/17sPsimk57wZ8OQehPn0pDWJFBboADv9xXBrnu-pfoL4 specifically we're thinking of adding base events (before any person/group/cohort processing and moving that to happen later in batches).

To be clear, nothing fundamental would change with the webhook/action matching. I mean, there will still be some code running for each event to check if it matches something or not? It'll still be performed on each event, but just without person/group/etc data now?

> Given that folks would like webhooks to be fast we'll want to move that to be on top of base events, so we don't have access to any person info. We might be forced to support the existing users by having those webhooks after the batching is done, but ...

From the standpoint of this HogVM implementation, there's little difference if person properties are present or not. All matching is done against an object you pass in. 

Cohort matching is now implemented as an async function that asks the plugin server for a true/false. I only implemented it to remain compatible with what we have now. Wouldn't have minded skipping, but here we are. Implementing cohort support did change the `execute` function's signature from `executeByteCode()` to `async executeByteCode()`, which might have implications further down the line. However I can easily make an optimised sync version even if we keep the async features (the sync version would return `{op,args,stack,i}` and would expect to be manually resumed with the answer, and/or throw).

> lets encourage users to not setup any new webhooks via hogQL that use persons or cohorts, not sure how to best do that though 😅 It would otherwise likely be very confusing to users why some webhooks are really slow and others fast

There are similar presentational problems with feature flags and cohorts. Feature flags do local evaluation on the client side, but against a `person` object. If a flag depends on a cohort that depends on a person property, it gets inlined into the matcher (I think?). If a flag depends on a behavioural cohort, it just won't match locally, and will come back later with the `/decide` response, maybe. Checking briefly, I didn't find anything special in the interface.

However for actions, I was thinking of showing a simple red line above a disabled "enable webhooks" button that just says something like: "Webhooks unavailable since the action accesses a cohort". If we do support a secondary path with person properties, we can say "you're in the slow lane now". I was anyway planning on adding an error if the action won't work if it uses forbidden HogQL functions, so expanding that is not a problem.

We can easily convert each action to HogQL, and we can also easily parse HogQL to see if it contains any cohort matching, person property access, etc. All of this in Python before saving the action.

tiina303 commented:
We already have HogQL to CH query, could we just use that? Seems like the simplest approach, especially with our move towards Data Warehouse land. Imagine we'll run a the CH query we get from you in batches against events table (or BaseEvents table) - we haven't finalised our plans on this front, but whatever we can we'll be doing in batches.

> If we do support a secondary path with person properties, we can say "you're in the slow lane now". I was anyway planning on adding an error if the action won't work if it uses forbidden HogQL functions, so expanding that is not a problem.

sounds great 👍

mariusandra commented:
> We already have HogQL to CH query, could we just use that? 

Well, that's not how action matching and webhooks currently work, so not really... or what do you mean?

We made the decision a while ago to do action matching live as data passes through the plugin server. Mainly for two reasons: 

1. I've never seen a "every 5 minutes fetch the last 5 minutes of data and do X" cron job run reliably. Never happened, can't happen, won't happen. We don't need to look far for examples: our own action matching used to work like this and was constantly falling behind. We periodically have issues with cohorts. Event volume calculation breaks every time we try. Don't get me started on insight and dashboard reloads. Based on being burned so many times, if we can avoid yet another "query the database because SQL is easy" solution, we absolutely should.
2. Querying for events after writing them into the database is wasteful. We have the event in memory. Instead of checking if it matches a few properties, we send it to a different server, let it write to disk, and then later make a bunch of queries to read it back from disk and make the same comparisons.

Those are the reasons why this was designed the way it was. If the architecture changes, I'm happy to revisit. It'll be very easy to swap this for a query, but I want to solve the problem of "hogql webhooks don't work" today.

mariusandra committed:
test for cohort matching

mariusandra committed:
it's an int

mariusandra committed:
not to run on every step

mariusandra committed:
can match against elements

mariusandra committed:
notify us if broken match

mariusandra commented:
@tiina303 does cohort matching for webhooks currently even work in the plugin server? I see that when matching we check the Postgres cohorts table... but that's empty for me locally. I tried to not break anything with this PR, but if this is altogether broken... then so much the better I guess.

mariusandra committed:
add bytecode_error

mariusandra committed:
add bytecode_error

mariusandra committed:
update package

mariusandra committed:
remove cohort test

mariusandra committed:
no bytecode

mariusandra commented:
I removed the cohort support, and added a warning into the interface when webhooks are disabled. See the gif in the issue description. 

You do need to save the action to see or remove the warning, which is a bit annoying. I'd like to defer fixing this to another PR... along with making the filters auto-reload the results below (see https://github.com/PostHog/posthog/issues/4505#issuecomment-1609670419).

mariusandra commented:
And one more thing... I currently implemented the [matching logic](https://github.com/PostHog/posthog/pull/16304/files#diff-297365893749599131f2948332c9cc1ecb46f8a0222d586bb779e2b20d474012R173-R207) as such:
1. If there's a bytecode and it matches, return true
2. Otherwise use the previous matching code
3. Send a message to sentry if the previous code returns a different error

Do we want this? Seems same to have it on at least in the beginning.

mariusandra committed:
several fixes

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
error field

mariusandra committed:
Merge branch 'hogvm-action-matching' of github.com:PostHog/posthog into hogvm-action-matching

mariusandra committed:
Merge branch 'master' into hogvm-action-matching

mariusandra committed:
fix tests

mariusandra committed:
fix tests again

mariusandra committed:
fix cohort tests

mariusandra committed:
update production dockerfile

mariusandra commented:
All right, test are green 🤞  and this is ready for a review!

mariusandra committed:
Merge branch 'master' into hogvm-action-matching

tiina303 commented:
> Well, that's not how action matching and webhooks currently work, so not really... or what do you mean?

When someone creates an action in the UI, then you show them the events that match that action (this is a CH query), I'm saying lets just use that.

> Querying for events after writing them into the database is wasteful. We have the event in memory. Instead of checking if it matches a few properties, we send it to a different server, let it write to disk, and then later make a bunch of queries to read it back from disk and make the same comparisons. That can't be better?

Currently what happens here is: we write events to kafka for CH to consume, we also read that Kafka topic as a whole in a separate pod async worker - which means loading all the data into the pod. Instead if we would look at the data in CH and only do something with the data that we need to send webhooks for, which is like 0.001% of events or less, then we'd be loading / processing way less data in total.

tiina303 commented:
If we look at only cohorts in PG not in CH, then we're not matching against dynamic cohorts cc @neilkakkar

mariusandra commented:
> > Well, that's not how action matching and webhooks currently work, so not really... or what do you mean?
> 
> When someone creates an action in the UI, then you show them the events that match that action (this is a CH query), I'm saying lets just use that.

Yes, but that's not how Webhooks work right now. Completely rearchitecting action matching is really out of the scope of this PR. I just want to get HogQL filters working and move on. I also said above I think this would be _much much much_ less reliable, and slower. You'd have to run at least one bespoke time-bounded lookback query _per team_ every $interval. Scheduling so many so frequent SQL queries sounds like a maintainability nightmare, and all that orchestration sounds like so much more work than what we're doing now. 

This sounds overall worse, so I'm missing something. What's the reason you're pushing for it? Is this blocking this PR?

> Currently what happens here is: we write events to kafka for CH to consume, we also read that Kafka topic as a whole in a separate pod async worker - which means loading all the data into the pod. Instead if we would look at the data in CH and only do something with the data that we need to send webhooks for, which is like 0.001% of events or less, then we'd be loading / processing way less data in total.

Ah, that makes sense, but we also run other plugins, right? E.g. GeoIP is run on most events (at least it once was? didn't look up stats now), and we should still allow [things like these](https://github.com/netdata/posthog-netdata-event-processing) to exist? Thus the data would anyway be in the server, we can just run a quick "path-so-hot-it's-compiled-by-the-V8-optimizer" synchronous instruction on the event, and call it a day?

In any case, if/when things change and querying becomes the way to go, we can switch this around very easily. Still, I'm worried about spawning 1k? 10k? 100k? queries every minute to check if a webhook matched. Cohorts can take time to calculate, but webhooks need to be instant for users to derive value from them.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

tiina303 commented:
> Ah, that makes sense, but we also run other plugins, right? E.g. GeoIP is run on most events (at least it once was? didn't look up stats now), and we should still allow [things like these](https://github.com/netdata/posthog-netdata-event-processing) to exist? Thus the data would anyway be in the server, we can just run a quick "path-so-hot-it's-compiled-by-the-V8-optimizer" synchronous instruction on the event, and call it a day?

geoIP is process_event <- that's ran in plugins-ingestion, webhooks and onEvent are run in ingestion-async (and I'm actually breaking them apart separate pods to be handling them). So no the data won't be on the server anyway.

Talked with the team and @timgl  - who will look into how people use webhooks first to understand better the requirements and what we want to build longer term.

timgl commented:
Let's separate out a couple of things going on here:
- Future of webhooks: I'll go and do some work to figure out what this should look like with the new pipeline plans. Otherwise not super relevant to this PR
- Should we merge this PR right now: given this is on the hot path of ingestion, Marius is about to leave for a week and the pipeline team would be on call for this and is already over stretched, I suggest we hold off on merging this until Marius is back and he can manage the roll out/get paged if things go wrong.

Does that work for everyone?

mariusandra commented:
@tiina303 I still haven't gotten a reply if there's any fundamental reason to not move on with this. Is there, or can I water this PR and try to get it merged?

tiina303 commented:
It make sense for me that Product Analytics owns the actions and matching part E2E (allowing you to provide a better user experience integrating views in the frontend), in which case you'd be quite free to do whatever and if it's broken or too slow I'd expect you to handle fixing it. 

I'd love for the interface to be `[url, body] = get getWebhooksToSend(db.pg, events)` you'd also do the message composing part there.

We should add some prom monitoring for timing around the whole actions matching thing and make sure we catch the exceptions.

mariusandra commented:
Thanks! Happy to own webhooks, and the related alerts, etc. I'll clean it PR up when I find time in this support fortnight I find myself in.

---

Also, for what it's worth, I do think there's a way to run all plugins sync and VM2-free by creating a new language called Hog, and rewriting all plugins in it (and also using it in notebooks). Using JavaScript for plugins with its unpredicatable async and security holes seems retroactively to have been a mistake. Our own language, with explicitly batched and routed async calls, may provide some very interesting benefits. Basically think of the Erlang VM running over Kafka. I could expand, but the idea is way too raw, and I don't feel like spending hours writing and arguing in RFCs :).

mariusandra committed:
Merge branch 'master' into hogvm-action-matching

mariusandra committed:
fix unrelated test

mariusandra commented on ee/clickhouse/queries/experiments/test_utils.py:
```diff
@@ -54,7 +54,11 @@ def test_with_feature_flag_properties_on_events(self):
+        ActionStep.objects.create(
+            action=action_credit_card,
+            event="paid",
+            properties=[{"key": "$os", "type": "event", "value": ["Windows"], "operator": "exact"}],
```
I checked and we're not using this oldschool style of properties anywhere, except these tests.

mariusandra replied:
I ran a query looking for such fields in the dashboard item filters, and found nothing. This was a while ago, so I lost the query 😬. Can try running again and posting results here.

mariusandra commented on posthog/temporal/tests/batch_exports/test_s3_batch_export_workflow.py:
```diff
@@ -598,6 +598,7 @@ async def test_s3_export_workflow_with_minio_bucket_and_custom_key_prefix(client
             "event": "test",
             "timestamp": "2023-04-25 13:30:00.000000",
             "created_at": "2023-04-25 13:30:00.000000",
+            "inserted_at": "2023-04-25 13:31:00.000000",
```
Tests were failing without this 🤷

Twixes committed:
Signal bytecode error on action page in subtler way

Twixes reviewed (changes requested):
The new system is great, my main concern is just that I think it's worth being cautious with rolling it out. Had fixed my fair share of problems with webhooks back in the day 😄

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
fix changelog

mariusandra commented on hogvm/CHANGELOG.md:
````diff
@@ -1,15 +1,6 @@
-```
-
-## 2023-06-28 - First version
+## 2023-06-30 - 1.0.2 - First version
````
Well then... I added this sad history back for all the world to see.

github-actions[bot] committed:
Update query snapshots

mariusandra committed:
split match for old and new

mariusandra committed:
Merge branch 'hogvm-action-matching' of github.com:PostHog/posthog into hogvm-action-matching

mariusandra committed:
attr_class handling

mariusandra commented on plugin-server/src/worker/ingestion/action-matcher.ts:
```diff
@@ -170,10 +171,46 @@ export class ActionMatcher {
+            return true
+        }
+
+        // If bytecode matching failed, fallback to previous matching
```
Makes sense. I tweaked the approach. Is this better?

If we merge this, then in the beginning it'll return a lot of false positives, as none of the actions will have the bytecode field populated. I'll run a separate async job to make sure they're all filled up. Running this as part of a migration feels risky.

Twixes replied:
Yup, this is great

mariusandra committed:
clean old stuff

mariusandra commented on posthog/api/action.py:
```diff
@@ -144,6 +149,8 @@ def update(self, instance: Any, validated_data: Dict[str, Any]) -> Any:
                         **{key: value for key, value in step.items() if key not in ("isNew", "selection")},
                     )
 
+        # bytecode might have been altered in the action steps
+        instance.refresh_from_db(fields=["bytecode", "bytecode_error"])
         instance = super().update(instance, validated_data)
```
Without this, it overrides the bytecode with an empty value. I might be doing something dum as well.

Twixes replied:
Ah, I see, because the `ActionStep` signal handler doesn't know about this `Action` instance

Twixes commented on plugin-server/src/worker/ingestion/action-matcher.ts:
```diff
@@ -137,16 +138,55 @@ export class ActionMatcher {
+                matches.push(teamActions[i])
+            }
+        }
+        this.statsd?.timing('action_matching_for_event_hogvm', matchingStart)
+        this.statsd?.increment('action_matches_found_hogvm', matches.length)
```
Isn't timing functionally a superset of increment? As in, I believe we can count the number of `action_matching_for_event_hogvm`.

Also, it's interesting to have the timing data for bytecode, but it would be _most_ interesting to compare bytecode and legacy matching, which requires analogous instrumentation in `checkActionLegacy`

Twixes commented on plugin-server/src/worker/ingestion/action-matcher.ts:
```diff
@@ -137,16 +138,55 @@ export class ActionMatcher {
+
+        if (
+            legacyResponse.length !== hogVMResponse.length ||
+            !legacyResponse.every((action, index) => action.id == hogVMResponse[index]?.id)
```
Is the order of actions guaranteed to be the same for both matching variants? Just to be safe it could be worth sorting the action IDs first

mariusandra replied:
I swapped the matches around:

```ts
const hogVMResponse = this.matchBytecode(event, elements)
const legacyResponse = await this.matchLegacy(event, elements) // this one was first
```

Now both of them call

```ts
const teamActions: Action[] = Object.values(this.actionManager.getTeamActions(event.teamId))
```

in succession before the await has a chance to mess up the order. Thus I'm reasonably sure the order will be the same.

Twixes commented on plugin-server/src/worker/ingestion/action-matcher.ts:
```diff
@@ -137,16 +138,55 @@ export class ActionMatcher {
+        } else {
+            this.statsd?.increment('action_matches_hogvm_legacy_match')
+        }
+        console.log({ hogVM: hogVMResponse.map((a) => a?.id), legacy: legacyResponse.map((a) => a?.id) })
```
Stray log

Twixes commented on plugin-server/src/worker/ingestion/action-matcher.ts:
```diff
@@ -137,16 +138,55 @@ export class ActionMatcher {
+
+    /** Get all actions matched to the event. */
+    public matchBytecode(event: PostIngestionEvent, elements?: Element[]): (Action | undefined)[] {
+        const matchingStart = new Date()
```
Let's use `performance.now()` instead of `new Date()`. We'll need to pass in `performance.now() - matchingStart` instead of just `matchingStart`, but `performance` is more reliable for metrics

Twixes commented on plugin-server/src/worker/ingestion/action-matcher.ts:
```diff
@@ -137,16 +138,55 @@ export class ActionMatcher {
+            legacyResponse.length !== hogVMResponse.length ||
+            !legacyResponse.every((action, index) => action.id == hogVMResponse[index]?.id)
+        ) {
+            this.statsd?.increment('action_matches_hogvm_legacy_mismatch')
+        } else {
+            this.statsd?.increment('action_matches_hogvm_legacy_match')
```
This is a case for tags, which would make it easier to break down by the possible cases. We could have a single metric like `this.statsd?.increment('action_matching_completed', { hogvm_delta: Math.sign(hogVMResponse.length - legacyResponse.length) })` (1 means false positive, -1 means false negative, 0 means matching results)

Twixes reviewed (commented):
A few comments on metrics instrumentation

mariusandra committed:
Merge branch 'master' into hogvm-action-matching

mariusandra committed:
more accurate matching

mariusandra committed:
debork merge

mariusandra committed:
debork merge 2

mariusandra commented on posthog/models/action_step.py:
```diff
@@ -50,13 +50,15 @@ class ActionStep(models.Model):
 
 @receiver(post_save, sender=ActionStep)
 def action_step_saved(sender, instance: ActionStep, created, **kwargs):
+    instance.action.refresh_bytecode()
```
Agreed, but I'm also unsure if we can do this better. Theoretically the bytecode is always in sync...

mariusandra committed:
cleaner css

mariusandra committed:
swap the matching order to hopefully eliminate side effects

mariusandra committed:
fix checkboxes

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update query snapshots

mariusandra commented on plugin-server/src/worker/ingestion/action-matcher.ts:
```diff
@@ -160,20 +200,61 @@ export class ActionMatcher {
+    ): boolean | undefined {
+        try {
+            if (Array.isArray(action.bytecode) && action.bytecode.length > 1) {
+                const eventProxy = new Proxy(event, {
```
I did this way because `elementsToString` feels like a complex operation, which we can opt out of performing on this hot path, unless something needs it.

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

mariusandra commented:
This PR has been split into:

https://github.com/PostHog/posthog/pull/16929
https://github.com/PostHog/posthog/pull/16930
https://github.com/PostHog/posthog/pull/16931
https://github.com/PostHog/posthog/pull/16932
https://github.com/PostHog/posthog/pull/16933
https://github.com/PostHog/posthog/pull/16937

Thus closing.
